Added Excel reading, database handling, and scheduling functionality

AmitHegde3 opened the pull request:
**Features Added:**
- Spring Boot Project Setup: Created a new Spring Boot project with dependencies for PostgreSQL, Apache POI, and Spring Scheduler.
- PostgreSQL Database Configuration: Configured the database connection in the application.properties file.
- Model and Repository Classes: Defined an entity class to map Excel data and created a repository interface for database interaction.
- Excel Service: Implemented a service to read data from the Excel sheet and perform insert/update operations in the database.
- Spring Scheduler Setup: Configured a cron job to periodically check for updates in the Excel sheet and update the database accordingly.(1 hr)
- Database Initialization: Ensured that the database is initialized with Excel data upon application startup.

AmitHegde3 committed:
Added Excel reading, database handling, and scheduling functionality

Features Added:
- Spring Boot Project Setup: Created a new Spring Boot project with dependencies for PostgreSQL, Apache POI, and Spring Scheduler.
- PostgreSQL Database Configuration: Configured the database connection in the application.properties file.
- Entity and Repository Classes: Defined an entity class to map Excel data and created a repository interface for database interaction.
- Excel Service: Implemented a service to read data from the Excel sheet and perform insert/update operations in the database.
- Spring Scheduler Setup: Configured a cron job to periodically check for updates in the Excel sheet and update the database accordingly.
- Database Initialization: Ensured that the database is initialized with Excel data upon application startup.

NabeelaFMS commented on src/main/java/com/sqltest/demo/DemoApplication.java:
Usaully we will name the SpringBootApplication class with Application name. eg: FmsApplication.java

AmitHegde3 replied:
Changed the class name to FmsApplication.java

NabeelaFMS commented on src/main/java/com/sqltest/demo/config/DataInitializer.java:
```diff
@@ -0,0 +1,23 @@
+        try {
+            excelService.readExcelAndSaveToDatabase();
+        } catch (IOException e) {
+            e.printStackTrace();
```
Introduce logger and add proper exception message instead of printStackTrace.
logger.info("Message", e.getMessage());

NabeelaFMS commented on src/main/java/com/sqltest/demo/config/ScheduledTasks.java:
```diff
@@ -0,0 +1,19 @@
+    @Autowired
+    private ExcelService excelService;
+
+    @Scheduled(cron = "0 0 * * * ?") // Run every hour
```
it will be costly if we run the schedulers in every hour, So define two times in a day. Specify a business time for FMS, Like in FMS application new intents are fetching only morning 10am and 4pm. Also add the cron value in application yaml and fetch the value in class.
eg: 
fms: 
  sync:
    cron: 0 7 0 * * ?

@Scheduled(cron = "${fms.sync.cron}")

AmitHegde3 replied:
Update code:
fms.sync.cron=0 0 10,16 * * ?(At 10am, 4pm)

NabeelaFMS commented on src/main/java/com/sqltest/demo/config/ScheduledTasks.java:
```diff
@@ -0,0 +1,19 @@
+    private ExcelService excelService;
+
+    @Scheduled(cron = "0 0 * * * ?") // Run every hour
+    public void checkAndUpdateDatabase() throws IOException {
```
dont throws exception, try to catch exception with loggers.

NabeelaFMS commented on src/main/java/com/sqltest/demo/model/SalesOrder.java:
```diff
@@ -0,0 +1,150 @@
+
+@Entity
+public class SalesOrder {
+//    @Id
```
remove commented codes

NabeelaFMS commented on src/main/java/com/sqltest/demo/model/SalesOrder.java:
```diff
@@ -0,0 +1,150 @@
+
+import java.util.List;
+
+@Entity
```
Remove all setter methods and add as annotion.
eg:
@Entity
@Table(name = "eample_table")
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public class Example{

AmitHegde3 replied:
All setters are removed and **Data** annotation is added  to generate getters, setters.

NabeelaFMS commented on src/main/java/com/sqltest/demo/service/ExcelService.java:
```diff
@@ -0,0 +1,119 @@
+    private static final String PLACEHOLDER = "N/A";
+
+    public void readExcelAndSaveToDatabase() throws IOException {
+        ClassPathResource resource = new ClassPathResource("excel/data.xlsx");
```
Define a specific path for the excel in application property file and fetch in class.
eg:
fms-so-dump: /fmsdir/SODUMP/

@Value("${fms-so-dump}")
	private String soDir;

AmitHegde3 replied:
Since the Excel is in the resources folder, I have used  ResourceLoader to load it from the classpath.

Updated code:
private static final String FILE_NAME = "data.xlsx";

 Resource resource = resourceLoader.getResource("classpath:excel/" + FILE_NAME);

NabeelaFMS commented on src/main/java/com/sqltest/demo/service/ExcelService.java:
```diff
@@ -0,0 +1,119 @@
+    private String getCellValueAsString(Cell cell) {
+        if (cell == null || cell.getCellType() == CellType.BLANK) {
+            return null;
+        } else {
```
when ever the if block comes true it will return from method. Then why a else {} needed here. when If block comes false automatically next statements will execute.

Refactor the code written in same way

NabeelaFMS commented on src/test/java/com/sqltest/demo/DemoApplicationTests.java:
please add unit testcases for all business logic classes.

AmitHegde3 replied:
**To be done.**

AmitHegde3 committed:
Update Fixes.

AmitHegde3 committed:
Update excel file path.

Defined the Excel file path using application.properties and @Value annotation in Spring Boot.
